File: chapter04/cpp/getRandom.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include "limits.h"
#include <math.h>
#include <stdlib.h>

using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int v=0, ListNode* nxt=nullptr) : val(v), next(nxt) {};
};

int get_random(ListNode* head)
{   
    srand((unsigned)time(NULL)); 
    ListNode* node = head->next;
    int choice = head->val;
    int k = 2;
    while(node != nullptr)
    {
        int r = rand() % k;
        cout << r << endl;
        if(r == 0)
            choice = node->val;

        node = node->next;
        k++;
    }
    return choice;
}

int main()
{
    ListNode* head = new ListNode(1);
    ListNode* tmp = head;

    for(int i=2; i<20; i++)
    {
        tmp->next = new ListNode(i);
        tmp = tmp->next;
    }

    cout << get_random(head) << endl;
    return 0;
}

File: chapter04/python/GetRandom.py
"""
问题: 给定一个长度未知的链表, 设计一个算法, 只能遍历一次, 均匀随机的返回链表中的一个节点
也就是每个节点被选中的概率为1/n

先说结论(很容易证明)，当你遇到第 i 个元素时，应该有 1/i 的概率选择该元素，1 - 1/i 的概率保持原有的选择。

同理，如果要随机选择 k 个数，只要在第 i 个元素处以 k/i 的概率选择该元素，以 1 - k/i 的概率保持原有选择即可
"""
import numpy as np


class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None


def get_random(head):
    p = head
    i = 0
    res = 0
    while p is not None:
        # 生成[0, i]之间的随机数
        r = np.random.randint(0, i + 1)

        if r == 0:
            res = p.val
        p = p.next
        i += 1
    return res


def get_random_k(head, k=1):
    p = head
    i = k
    res = [0] * k
    for i in range(k):
        res[i] = p.val
        p = p.next

    while p is not None:
        # 生成[0, i)之间的随机数
        r = np.random.randint(0, i + 1)

        if r < k:
            res[r] = p.val
        p = p.next
        i += 1
    return res


if __name__ == "__main__":
    head = ListNode()

    tmp = head
    for i in range(1, 5):
        tmp.next = ListNode(i)
        tmp = tmp.next
    print(get_random_k(head, 2))

